File: dart/dartsearch/lib/src/searcher.dart
import 'dart:async';
import 'dart:io';
import 'dart:convert';
import 'dart:math';

import 'package:dartfind/dartfind.dart';

import 'package:dartsearch/src/search_exception.dart';
import 'package:dartsearch/src/search_result.dart';
import 'package:dartsearch/src/search_settings.dart';
// import 'package:path/path.dart' as path;

class Searcher {
  final SearchSettings settings;
  late final FileTypes _fileTypes;
  late final Finder _finder;
  // whether to allow invalid/malformed characters in encoding/decoding
  // setting to false means search will end once an attempt is made to read
  // a file in an incompatible encoding
  final bool allowInvalid = true;
  late final Encoding _binaryEncoding;
  Encoding? _textEncoding = systemEncoding;

  Searcher(this.settings) {
    _fileTypes = FileTypes();
    try {
      _finder = Finder(settings);
    } on FindException catch (e) {
      throw SearchException(e.message);
    } catch (e) {
      throw SearchException(
          'Unknown error occurred trying to instantiate finder');
    }
    _validateSettings();
    _binaryEncoding = Latin1Codec(allowInvalid: allowInvalid);
  }

  void _validateSettings() {
    if (settings.searchPatterns.isEmpty) {
      throw SearchException('No search patterns defined');
    }
    if (settings.linesAfter < 0) {
      throw SearchException('Invalid linesafter');
    }
    if (settings.linesBefore < 0) {
      throw SearchException('Invalid linesbefore');
    }
    if (settings.maxLineLength < 0) {
      throw SearchException('Invalid maxlinelength');
    }
    _textEncoding = Encoding.getByName(settings.textFileEncoding)!;
    if (_textEncoding == null) {
      throw SearchException(
          'Invalid or unsupported encoding: ${settings.textFileEncoding}');
    } else if (allowInvalid) {
      if (_textEncoding == utf8) {
        _textEncoding = Utf8Codec(allowMalformed: true);
      } else if (_textEncoding == latin1) {
        _textEncoding = Latin1Codec(allowInvalid: true);
      } else if (_textEncoding == ascii) {
        _textEncoding = AsciiCodec(allowInvalid: true);
      }
    }
  }

  bool _anyMatchesAnyPattern(Iterable<String> elems, Set<Pattern> patterns) {
    return elems.any((elem) => _matchesAnyPattern(elem, patterns));
  }

  bool _matchesAnyPattern(String s, Set<Pattern> patterns) {
    return patterns.any((p) => p.allMatches(s).isNotEmpty);
  }

  Future<List<SearchResult>> _searchBinaryFile(FileResult fr) async {
    if (settings.debug) {
      logMsg('Searching binary file $fr');
    }
    return fr.file
        .readAsString(encoding: _binaryEncoding)
        .then((String contents) {
      var results = <SearchResult>[];
      for (var p in settings.searchPatterns) {
        var matches = [];
        if (settings.firstMatch) {
          var match = (p as RegExp).firstMatch(contents);
          if (match != null) {
            matches = [match];
          }
        } else {
          matches = p.allMatches(contents).toList();
        }
        for (var m in matches) {
          results.add(
              SearchResult(p, fr, 0, m.start + 1, m.end + 1, null, [], []));
        }
      }
      return results;
    });
  }

  bool _linesMatch(
      List<String> lines, Set<Pattern> inPatterns, Set<Pattern> outPatterns) {
    return lines.isEmpty ||
        ((inPatterns.isEmpty || _anyMatchesAnyPattern(lines, inPatterns)) &&
            (outPatterns.isEmpty ||
                !_anyMatchesAnyPattern(lines, outPatterns)));
  }

  Future<List<SearchResult>> searchLineStream(Stream<String> stream) async {
    var results = <SearchResult>[];
    var lineNum = 0;
    var matchedPatterns = {};
    if (settings.firstMatch) {
      stream = stream.takeWhile((_) {
        return matchedPatterns.length < settings.searchPatterns.length;
      });
    }

    try {
      var linesBefore = <String>[];
      var linesAfter = <String>[];

      var it = StreamIterator(stream);

      while (true) {
        lineNum++;
        String line;
        if (linesAfter.isNotEmpty) {
          line = linesAfter.removeAt(0);
        } else if (await it.moveNext()) {
          line = it.current;
        } else {
          break;
        }
        if (settings.linesAfter > 0) {
          while (
              linesAfter.length < settings.linesAfter && await it.moveNext()) {
            linesAfter.add(it.current);
          }
        }

        var searchPatterns = <Pattern>{};
        if (settings.firstMatch) {
          searchPatterns = settings.searchPatterns
              .where((elem) => !matchedPatterns.containsKey(elem))
              .toSet();
        } else {
          searchPatterns = settings.searchPatterns;
        }
        if (searchPatterns.isEmpty) {
          break;
        }
        for (var p in searchPatterns) {
          var matches = [];
          if (settings.firstMatch) {
            var match = (p as RegExp).firstMatch(line);
            if (match != null) {
              matches = [match];
              matchedPatterns[p] = 1;
            }
          } else {
            matches = p.allMatches(line).toList();
          }

          if (matches.isEmpty ||
              !_linesMatch(linesBefore, settings.inLinesBeforePatterns,
                  settings.outLinesBeforePatterns) ||
              !_linesMatch(linesAfter, settings.inLinesAfterPatterns,
                  settings.outLinesAfterPatterns)) {
            continue;
          }

          for (var m in matches) {
            results.add(SearchResult(p, null, lineNum, m.start + 1, m.end + 1,
                line, linesBefore.toList(), linesAfter.toList()));
          }
        }

        if (settings.linesBefore > 0) {
          if (linesBefore.length == settings.linesBefore) {
            linesBefore.removeAt(0);
          }
          if (linesBefore.length < settings.linesBefore) {
            linesBefore.add(line);
          }
        }
        if (settings.firstMatch &&
            matchedPatterns.length == settings.searchPatterns.length) {
          break;
        }
      }

      return results;
    } catch (e) {
      logMsg(e.toString());
      return [];
    }
  }

  Future<List<SearchResult>> _searchTextFileLines(FileResult fr) async {
    var results = <SearchResult>[];
    var inputStream = fr.file.openRead();
    try {
      var lineStream =
          _textEncoding!.decoder.bind(inputStream).transform(LineSplitter());
      return searchLineStream(lineStream).then((results) {
        return results.map((r) {
          r.file = fr;
          return r;
        }).toList();
      });
    } on FormatException catch (e) {
      logMsg('Error reading ${fr.file}: ${e.message}');
    } catch (e) {
      logMsg(e.toString());
    }
    return results;
  }

  List<String> _getLinesFromMultilineString(
      String s, List<int> newLineIndices) {
    var lines = <String>[];
    for (var i = 0; i < newLineIndices.length - 1; ++i) {
      lines.add(s.substring(newLineIndices[i] + 1, newLineIndices[i + 1]));
    }
    return lines;
  }

  List<SearchResult> searchMultilineString(String s) {
    var results = <SearchResult>[];
    var newLineIndices = <int>[];
    var it = s.runes.iterator;
    var i = 0;
    while (it.moveNext()) {
      if (it.current == 10) {
        newLineIndices.add(i);
      }
      i++;
    }

    for (var p in settings.searchPatterns) {
      var matches = [];
      if (settings.firstMatch) {
        var match = (p as RegExp).firstMatch(s);
        if (match != null) {
          matches = [match];
        }
      } else {
        matches = p.allMatches(s).toList();
      }
      for (var m in matches) {
        var beforeNewlineIndices =
            newLineIndices.takeWhile((i) => i <= m.start).toList();
        var linesBefore = <String>[];
        if (settings.linesBefore > 0) {
          var linesBeforeIndices = beforeNewlineIndices.reversed
              .take(settings.linesBefore + 1)
              .toList()
              .reversed
              .toList();
          linesBefore = _getLinesFromMultilineString(s, linesBeforeIndices);
        }
        var afterNewlineIndices =
            newLineIndices.skipWhile((i) => i <= m.start).toList();
        var linesAfter = <String>[];
        if (settings.linesAfter > 0) {
          var linesAfterIndices =
              afterNewlineIndices.take(settings.linesAfter + 1).toList();
          linesAfter = _getLinesFromMultilineString(s, linesAfterIndices);
        }

        if (!_linesMatch(linesBefore, settings.inLinesBeforePatterns,
                settings.outLinesBeforePatterns) ||
            !_linesMatch(linesAfter, settings.inLinesAfterPatterns,
                settings.outLinesAfterPatterns)) {
          continue;
        }

        var lineNum = beforeNewlineIndices.length + 1;
        var startLineIndex = 0;
        if (lineNum > 1) {
          startLineIndex = beforeNewlineIndices.last + 1;
        }
        var endLineIndex = s.length - 1;
        if (afterNewlineIndices.isNotEmpty) {
          endLineIndex = afterNewlineIndices.first;
        }
        var line = s.substring(startLineIndex, endLineIndex);
        var startMatchIndex = m.start - startLineIndex + 1;
        var endMatchIndex = m.end - startLineIndex + 1;
        results.add(SearchResult(p, null, lineNum, startMatchIndex,
            endMatchIndex, line, linesBefore, linesAfter));
      }
    }

    return results;
  }

  Future<List<SearchResult>> _searchTextFileContents(FileResult fr) async {
    return fr.file
        .readAsString(encoding: _textEncoding!)
        .then((String contents) {
      var results = searchMultilineString(contents);
      return results.map((r) {
        r.file = fr;
        return r;
      }).toList();
    });
  }

  Future<List<SearchResult>> _searchTextFile(FileResult fr) async {
    if (settings.debug) {
      logMsg('Searching text file $fr');
    }
    if (settings.multiLineSearch) {
      return _searchTextFileContents(fr);
    } else {
      return _searchTextFileLines(fr);
    }
  }

  Future<List<SearchResult>> _searchFile(FileResult fr) async {
    var results = <SearchResult>[];
    if ({FileType.text, FileType.code, FileType.xml}.contains(fr.fileType)) {
      return _searchTextFile(fr);
    } else if (fr.fileType == FileType.binary) {
      return _searchBinaryFile(fr);
    }
    return results;
  }

  int cmpByMatchLocation(SearchResult r1, SearchResult r2) {
    if (r1.lineNum != r2.lineNum) {
      return r1.lineNum.compareTo(r2.lineNum);
    }
    if (r1.matchStartIndex != r2.matchStartIndex) {
      return r1.matchStartIndex.compareTo(r2.matchStartIndex);
    }
    return r1.matchEndIndex.compareTo(r2.matchEndIndex);
  }

  int cmpByFilePath(SearchResult r1, SearchResult r2) {
    var cmp = 0;
    if (r1.file != null && r2.file != null) {
      cmp = _finder.cmpByFilePath(r1.file!, r2.file!);
    }
    if (cmp != 0) {
      return cmp;
    }
    return cmpByMatchLocation(r1, r2);
  }

  int cmpByFileName(SearchResult r1, SearchResult r2) {
    var cmp = 0;
    if (r1.file != null && r2.file != null) {
      cmp = _finder.cmpByFileName(r1.file!, r2.file!);
    }
    if (cmp != 0) {
      return cmp;
    }
    return cmpByMatchLocation(r1, r2);
  }

  int cmpByFileSize(SearchResult r1, SearchResult r2) {
    var cmp = 0;
    if (r1.file != null && r2.file != null) {
      cmp = _finder.cmpByFileSize(r1.file!, r2.file!);
    }
    if (cmp != 0) {
      return cmp;
    }
    return cmpByMatchLocation(r1, r2);
  }

  int cmpByFileType(SearchResult r1, SearchResult r2) {
    var cmp = 0;
    if (r1.file != null && r2.file != null) {
      cmp = _finder.cmpByFileType(r1.file!, r2.file!);
    }
    if (cmp != 0) {
      return cmp;
    }
    return cmpByMatchLocation(r1, r2);
  }

  int cmpByLastMod(SearchResult r1, SearchResult r2) {
    var cmp = 0;
    if (r1.file != null && r2.file != null) {
      cmp = _finder.cmpByLastMod(r1.file!, r2.file!);
    }
    if (cmp != 0) {
      return cmp;
    }
    return cmpByMatchLocation(r1, r2);
  }

  int Function(SearchResult, SearchResult)? getSortComparator() {
    if (settings.sortDescending) {
      switch (settings.sortBy) {
        case SortBy.fileName:
          return (SearchResult r1, SearchResult r2) => cmpByFileName(r2, r1);
        case SortBy.fileSize:
          return (SearchResult r1, SearchResult r2) => cmpByFileSize(r2, r1);
        case SortBy.fileType:
          return (SearchResult r1, SearchResult r2) => cmpByFileType(r2, r1);
        case SortBy.lastMod:
          return (SearchResult r1, SearchResult r2) => cmpByLastMod(r2, r1);
        default:
          return (SearchResult r1, SearchResult r2) => cmpByFilePath(r2, r1);
      }
    }
    switch (settings.sortBy) {
      case SortBy.fileName:
        return (SearchResult r1, SearchResult r2) => cmpByFileName(r1, r2);
      case SortBy.fileSize:
        return (SearchResult r1, SearchResult r2) => cmpByFileSize(r1, r2);
      case SortBy.fileType:
        return (SearchResult r1, SearchResult r2) => cmpByFileType(r1, r2);
      case SortBy.lastMod:
        return (SearchResult r1, SearchResult r2) => cmpByLastMod(r1, r2);
      default:
        return (SearchResult r1, SearchResult r2) => cmpByFilePath(r1, r2);
    }
  }

  void sortSearchResults(List<SearchResult> results) {
    var sortComparator = getSortComparator();
    results.sort(sortComparator);
  }

  Future<List<SearchResult>> _searchFiles() async {
    var fileResults = await _finder.find();
    if (settings.verbose) {
      var searchDirs =
          fileResults.map((fr) => fr.file.parent.path).toSet().toList();
      logMsg('\nDirectories to be searched (${searchDirs.length}):');
      for (var d in searchDirs) {
        logMsg(FileUtil.contractPath(d));
      }
      logMsg('\nFiles to be searched (${fileResults.length}):');
      for (var fr in fileResults) {
        logMsg(FileUtil.contractPath(fr.file.path));
      }
    }
    // this is the (almost) largest batch size you can have before you get the
    // "too many files open" errors
    var batchSize = 245;
    var offset = 0;

    var results = <SearchResult>[];

    while (offset < fileResults.length) {
      var toIndex = min(offset + batchSize, fileResults.length);
      var fileResultsFutures =
          fileResults.sublist(offset, toIndex).map((sf) => _searchFile(sf));
      await Future.wait(fileResultsFutures).then((filesResults) {
        for (var fileResults in filesResults) {
          results.addAll(fileResults);
        }
      });
      offset += batchSize;
    }

    if (results.length > 1) {
      sortSearchResults(results);
    }
    return results;
  }

  Future<List<SearchResult>> search() async {
    return Future.wait([_fileTypes.ready]).then((res) {
      return _searchFiles();
    });
  }

  //////////////////////////////////////////////////////////////////////////////

  void printResults(
      List<SearchResult> results, SearchResultFormatter formatter) {
    if (results.isNotEmpty) {
      logMsg('\nSearch results (${results.length}):');
      for (var r in results) {
        logMsg(formatter.format(r));
      }
    } else {
      logMsg('\nSearch results: 0');
    }
  }

  int sortFileResults(FileResult fr1, FileResult fr2) {
    if (fr1.file.parent.path == fr2.file.parent.path) {
      return fr1.file.path.compareTo(fr2.file.path);
    } else {
      return fr1.file.parent.path.compareTo(fr2.file.parent.path);
    }
  }

  List<FileResult> getMatchingFileResults(List<SearchResult> results) {
    var files = results.map((r) => r.file!).toSet().toList();
    files.sort(sortFileResults);
    return files;
  }

  void printMatchingDirs(
      List<SearchResult> results, SearchResultFormatter formatter) {
    var files = getMatchingFileResults(results);
    _finder.printMatchingDirs(files, formatter.fileFormatter!);
  }

  void printMatchingFiles(
      List<SearchResult> results, SearchResultFormatter formatter) {
    var files = getMatchingFileResults(results);
    _finder.printMatchingFiles(files, formatter.fileFormatter!);
  }

  List<String> getMatchingLines(
      List<SearchResult> results, SearchSettings settings) {
    var lines = results
        .where((r) => r.line != null)
        .map((r) => r.line!.trim())
        .toList();
    if (settings.uniqueLines) {
      lines = lines.toSet().toList();
    }
    lines.sort((l1, l2) => l1.toLowerCase().compareTo(l2.toLowerCase()));
    return lines;
  }

  void printMatchingLines(
      List<SearchResult> results, SearchResultFormatter formatter) {
    var lines = getMatchingLines(results, settings);
    String msg;
    if (settings.uniqueLines) {
      msg = '\nUnique matching lines';
    } else {
      msg = '\nMatching lines';
    }
    if (lines.isNotEmpty) {
      logMsg('$msg (${lines.length}):');
      for (var line in lines) {
        logMsg(formatter.formatLine(line));
      }
    } else {
      logMsg('$msg: 0');
    }
  }
}

File: php/phpsearch/src/phpsearch/Searcher.php
<?php

declare(strict_types=1);

namespace phpsearch;

use phpfind\FileResult;
use phpfind\FileType;
use phpfind\Finder;
use phpfind\FindException;
use phpfind\Logger;

/**
 * Class Searcher
 */
class Searcher
{
    private readonly SearchSettings $settings;
    private readonly Finder $finder;

    /**
     * @throws SearchException
     */
    public function __construct(SearchSettings $settings)
    {
        $this->settings = $settings;
        try {
            $this->finder = new Finder($settings);
        } catch (FindException $e) {
            throw new SearchException($e->getMessage());
        }
        // We can validate_settings after creating Finder since we want its validation to occur first
        $this->validate_settings();
    }

    /**
     * @throws SearchException
     */
    private function validate_settings(): void
    {
        if (count($this->settings->search_patterns) == 0) {
            throw new SearchException('No search patterns defined');
        }
        if ($this->settings->lines_after < 0) {
            throw new SearchException('Invalid lines_after');
        }
        if ($this->settings->lines_before < 0) {
            throw new SearchException('Invalid lines_before');
        }
        if ($this->settings->max_line_length < 0) {
            throw new SearchException('Invalid max_line_length');
        }
    }

    private function matches_any_pattern(string $s, array $patterns): bool
    {
        foreach ($patterns as $pattern) {
            $pattern = '/' . $pattern . '/';
            if (preg_match_all($pattern, $s)) {
                return true;
            }
        }
        return false;
    }

    private function any_matches_any_pattern(array $slist, array $patterns): bool
    {
        foreach ($slist as $s) {
            if ($this->matches_any_pattern($s, $patterns)) {
                return true;
            }
        }
        return false;
    }

    private function get_new_line_indices(string $s): array
    {
        $indices = array();
        $i = 0;
        while ($i < strlen($s)) {
            if ($s[$i] == "\n") {
                $indices[] = $i;
            }
            $i++;
        }
        return $indices;
    }

    private function get_before_indices(array $indices, int $after_index): array
    {
        $less_or_equal = function ($i) use ($after_index) {
            return $i <= $after_index;
        };
        return array_filter($indices, $less_or_equal);
    }

    private function get_after_indices(array $indices, int $before_index): array
    {
        $greater_than = function ($i) use ($before_index) {
            return $i > $before_index;
        };
        return array_filter($indices, $greater_than);
    }

    private function get_lines_before(
        string $s,
        array $before_start_indices,
        array $start_line_indices,
        array $end_line_indices
    ): array {
        $lines_before = array();
        $i = 0;
        while ($i < count($before_start_indices) && $i < $this->settings->lines_before) {
            $start_index = $before_start_indices[$i];
            $end_index = $end_line_indices[array_search($start_index, $start_line_indices)];
            $line = substr($s, $start_index, $end_index - $start_index);
            $lines_before[] = $line;
            $i++;
        }
        return $lines_before;
    }

    private function get_lines_after(
        string $s,
        array $after_start_indices,
        array $start_line_indices,
        array $end_line_indices
    ): array {
        $lines_after = array();
        $i = 0;
        while ($i < count($after_start_indices) && $i < $this->settings->lines_after) {
            $start_index = $after_start_indices[$i];
            $end_index = $end_line_indices[array_search($start_index, $start_line_indices)];
            $line = substr($s, $start_index, $end_index - $start_index);
            $lines_after[] = $line;
            $i++;
        }
        return $lines_after;
    }

    private function lines_match(array $lines, array $in_patterns, array $out_patterns): bool
    {
        if ((!$in_patterns || $this->any_matches_any_pattern($lines, $in_patterns))
            &&
            (!$out_patterns || !$this->any_matches_any_pattern($lines, $out_patterns))) {
            return true;
        }
        return false;
    }

    private function lines_before_match(array $lines_before): bool
    {
        return $this->lines_match(
            $lines_before,
            $this->settings->in_lines_before_patterns,
            $this->settings->out_lines_before_patterns
        );
    }

    private function lines_after_match(array $lines_after): bool
    {
        return $this->lines_match(
            $lines_after,
            $this->settings->in_lines_after_patterns,
            $this->settings->out_lines_after_patterns
        );
    }

    /**
     * @return SearchResult[]
     */
    public function search_multi_line_string(string $s): array
    {
        $results = array();
        $new_line_indices = $this->get_new_line_indices($s);
        $plus_one = function ($i) {
            return $i + 1;
        };
        $start_line_indices = array_merge(array(0), array_map($plus_one, $new_line_indices));
        $end_line_indices = array_merge($new_line_indices, array(strlen($s) - 1));
        foreach ($this->settings->search_patterns as $pattern) {
            $p = '/' . $pattern . '/';
            if (preg_match_all($p, $s, $matches, PREG_OFFSET_CAPTURE)) {
                foreach ($matches[0] as $m) {
                    $start_index = $m[1];
                    $end_index = $start_index + strlen($m[0]);
                    $before_start_indices = $this->get_before_indices($start_line_indices, $start_index);
                    $m_line_start_index = 0;
                    $before_line_count = 0;
                    if ($before_start_indices) {
                        $before_line_count = count($before_start_indices) - 1;
                        $m_line_start_index = $before_start_indices[$before_line_count];
                    }
                    $m_line_end_index = $end_line_indices[array_search($m_line_start_index, $start_line_indices)];
                    $line = substr($s, $m_line_start_index, $m_line_end_index - $m_line_start_index);
                    $lines_before = [];
                    if ($this->settings->lines_before) {
                        $lines_before = $this->get_lines_before(
                            $s,
                            array_slice(
                                $before_start_indices,
                                ($this->settings->lines_before + 1) * -1,
                                -1
                            ),
                            $start_line_indices,
                            $end_line_indices
                        );
                    }
                    $lines_after = [];
                    if ($this->settings->lines_after) {
                        $after_start_indices = $this->get_after_indices($start_line_indices, $start_index);
                        $lines_after = $this->get_lines_after(
                            $s,
                            array_slice(
                                $after_start_indices,
                                0,
                                $this->settings->lines_after
                            ),
                            $start_line_indices,
                            $end_line_indices
                        );
                    }
                    if (($lines_before && !$this->lines_before_match($lines_before))
                        ||
                        ($lines_after && !$this->lines_after_match($lines_after))) {
                        continue;
                    }
                    $r = new SearchResult(
                        $pattern,
                        null,
                        $before_line_count + 1,
                        $start_index - $m_line_start_index + 1,
                        $end_index - $m_line_start_index + 1,
                        $line,
                        $lines_before,
                        $lines_after
                    );
                    $results[] = $r;
                    if ($this->settings->first_match) {
                        break;
                    }
                }
            }
        }
        return $results;
    }

    /**
     * @return SearchResult[]
     */
    private function search_text_file_contents(FileResult $f): array
    {
        $contents = file_get_contents($f->file_path());
        $results = $this->search_multi_line_string($contents);
        foreach ($results as $r) {
            $r->file = $f;
        }
        return $results;
    }

    /**
     * @return SearchResult[]
     */
    public function search_lines(array $lines): array
    {
        $line_num = 0;
        $line = '';
        $lines_before = array();
        $lines_after = array();
        $pattern_match_map = array();
        $results = array();
        while (true) {
            if ($lines_after) {
                $line = array_shift($lines_after);
            } else {
                $line = array_shift($lines);
            }
            if ($line === null) {
                break;
            }
            $line_num++;
            if ($this->settings->lines_after) {
                while (count($lines_after) < $this->settings->lines_after) {
                    $line_after = array_shift($lines);
                    if ($line_after == null) {
                        break;
                    } else {
                        $lines_after[] = $line_after;
                    }
                }
            }
            foreach ($this->settings->search_patterns as $pattern) {
                if ($this->settings->first_match && array_key_exists($pattern, $pattern_match_map)) {
                    continue;
                }
                $p = '/' . $pattern . '/';
                if (preg_match_all($p, $line, $matches, PREG_OFFSET_CAPTURE)) {
                    if (($lines_before && !$this->lines_before_match($lines_before))
                        ||
                        ($lines_after && !$this->lines_after_match($lines_after))) {
                        continue;
                    }
                    foreach ($matches[0] as $m) {
                        $start_index = $m[1] + 1;
                        $end_index = $start_index + strlen($m[0]);
                        $r = new SearchResult(
                            $pattern,
                            null,
                            $line_num,
                            $start_index,
                            $end_index,
                            $line,
                            $lines_before,
                            $lines_after
                        );
                        $results[] = $r;
                        $pattern_match_map[$pattern] = 1;
                    }
                }
            }
            if ($this->settings->lines_before) {
                if (count($lines_before) == $this->settings->lines_before) {
                    array_shift($lines_before);
                }
                if (count($lines_before) < $this->settings->lines_before) {
                    $lines_before[] = $line;
                }
            }
        }
        return $results;
    }

    /**
     * @return SearchResult[]
     */
    private function search_text_file_lines(FileResult $f): array
    {
        $lines = file($f->file_path());
        $results = $this->search_lines($lines);
        foreach ($results as $r) {
            $r->file = $f;
        }
        return $results;
    }

    /**
     * @return SearchResult[]
     */
    private function search_text_file(FileResult $f): array
    {
        if ($this->settings->debug) {
            Logger::log_msg("Searching text file $f");
        }
        if ($this->settings->multi_line_search) {
            return $this->search_text_file_contents($f);
        } else {
            return $this->search_text_file_lines($f);
        }
    }

    /**
     * @return SearchResult[]
     */
    private function search_binary_file(FileResult $f): array
    {
        if ($this->settings->debug) {
            Logger::log_msg("Searching binary file $f");
        }
        $contents = file_get_contents($f->file_path());
        $results = [];
        foreach ($this->settings->search_patterns as $pattern) {
            $p = '/' . $pattern . '/';
            if (preg_match_all($p, $contents, $matches, PREG_OFFSET_CAPTURE)) {
                foreach ($matches[0] as $m) {
                    $start_index = $m[1] + 1;
                    $end_index = $start_index + strlen($m[0]);
                    $r = new SearchResult(
                        $pattern,
                        $f,
                        0,
                        $start_index,
                        $end_index,
                        null,
                        [],
                        []
                    );
                    $results[] = $r;
                    if ($this->settings->first_match) {
                        break;
                    }
                }
            }
        }
        return $results;
    }

    /**
     * @return SearchResult[]
     */
    public function search_file(FileResult $f): array
    {
        if ($f->file_type == FileType::Text || $f->file_type == FileType::Code ||
            $f->file_type == FileType::Xml) {
            return $this->search_text_file($f);
        } elseif ($f->file_type == FileType::Binary) {
            return $this->search_binary_file($f);
        }
        return [];
    }

    /**
     * @return SearchResult[]
     * @throws SearchException
     */
    public function search(): array
    {
        try {
            $file_results = $this->finder->find();
        } catch (FindException $e) {
            throw new SearchException($e->getMessage());
        }
        if ($this->settings->verbose) {
            $get_path = function ($sf) {
                return $sf->path;
            };
            $dir_results = array_unique(array_map($get_path, $file_results));
            Logger::log_msg(sprintf("\nDirectories to be searched (%d):", count($dir_results)));
            foreach ($dir_results as $d) {
                Logger::log_msg($d);
            }
            Logger::log_msg(sprintf("\n\nFiles to be searched (%d):", count($file_results)));
            foreach ($file_results as $f) {
                Logger::log_msg((string)$f);
            }
        }

        // search the files
        $search_results = [];
        foreach ($file_results as $f) {
            $search_results = array_merge($search_results, $this->search_file($f));
        }
        if (count($search_results) > 1) {
            $search_result_sorter = new SearchResultSorter($this->settings);
            return $search_result_sorter->sort($search_results);
        }
        return $search_results;
    }

    public function print_results(array $results, SearchResultFormatter $formatter): void
    {
        if (count($results) > 0) {
            Logger::log_msg(sprintf("\nSearch results (%d):", count($results)));
            foreach ($results as $r) {
                Logger::log_msg($formatter->format($r));
            }
        } else {
            Logger::log_msg("\nSearch results: 0");
        }
    }

    public function get_file_results(array $results): array
    {
        $file_hash = array();
        $file_results = array();
        foreach ($results as $r) {
            if (null != $r->file) {
                $f = $r->file;
                if (!array_key_exists($f->file_path(), $file_hash)) {
                    $file_results[] = $f;
                    $file_hash[$f->file_path()] = $f;
                }
            }
        }
        return $file_results;
    }

    public function print_matching_dirs(array $results, SearchResultFormatter $formatter): void
    {
        $file_results = $this->get_file_results($results);
        $this->finder->print_matching_dirs($file_results, $formatter->file_result_formatter);
    }

    public function print_matching_files(array $results, SearchResultFormatter $formatter): void
    {
        $file_results = $this->get_file_results($results);
        $this->finder->print_matching_files($file_results, $formatter->file_result_formatter);
    }

    public function get_matching_lines(array $results): array
    {
        $lines = array();
        foreach ($results as $r) {
            $l = trim($r->line);
            if (!$this->settings->unique_lines || !in_array($l, $lines)) {
                $lines[] = $l;
            }
        }
        if ($this->settings->sort_case_insensitive) {
            usort($lines, 'cmp_ignorecase');
        } else {
            sort($lines);
        }
        return $lines;
    }

    public function print_matching_lines(array $results, SearchResultFormatter $formatter): void
    {
        $lines = $this->get_matching_lines($results);
        $msg = "\nMatching lines";
        if ($this->settings->unique_lines) {
            $msg = "\nUnique matching lines";
        }
        if (count($lines) > 0) {
            Logger::log_msg(sprintf("%s (%d):", $msg, count($lines)));
            foreach ($lines as $l) {
                Logger::log_msg($formatter->format_line($l));
            }
        } else {
            Logger::log_msg(sprintf("%s: 0", $msg));
        }
    }

    public function get_matches(array $results): array
    {
        $matches = array();
        foreach ($results as $r) {
            $m = substr($r->line, $r->match_start_index - 1, $r->match_end_index - $r->match_start_index);
            if (!$this->settings->unique_lines || !in_array($m, $matches)) {
                $matches[] = $m;
            }
        }
        if ($this->settings->sort_case_insensitive) {
            usort($matches, 'cmp_ignorecase');
        } else {
            sort($matches);
        }
        return $matches;
    }

    public function print_matches(array $results, SearchResultFormatter $formatter): void
    {
        $matches = $this->get_matches($results);
        $msg = "\nMatches";
        if ($this->settings->unique_lines) {
            $msg = "\nUnique matches";
        }
        if (count($matches) > 0) {
            Logger::log_msg(sprintf("%s (%d):", $msg, count($matches)));
            foreach ($matches as $m) {
                Logger::log_msg($formatter->format_match($m));
            }
        } else {
            Logger::log_msg(sprintf("%s: 0", $msg));
        }
    }
}
